examples/rlhf/train_rlhf.py

```diff
@@ -0,0 +1,339 @@
+        batch_size=ppo_batch_size,
+        sampler=SamplerWithoutReplacement(),
+        prefetch=10,
+    )
+
+    rollout_from_model = RolloutFromModel(model, ref_model, reward_model)
+
+    best_val_reward = float("-inf")
+    it = 0  # it is equivalent to batch_size number of episodes
+    with tqdm(total=int(max_epochs * num_rollouts_per_epoch / batch_size)) as pbar:
+        for _epoch in range(1, max_epochs + 1):
+            rb.empty()
+            rollout_rewards = []
+            rollout_kl = []
+            kl_controller = AdaptiveKLController(0.1, 6, 10000)
+            for _ in range(0, num_rollouts_per_epoch, batch_size):
+                batch = next(train_loader)
+                td = rollout_from_model.rollout_from_data(
+                    batch, kl_coef=kl_controller.value
+                )
+                with torch.no_grad(), ctx:
+                    # moving this to within epoch
+                    adv_fn(td)
+                # it's possible we didn't fill the replay buffer in the last iteration if
+                # generation stopped early, so we empty first before repopulating
+                rb.extend(flatten_td(td))
+                done = td.get(("next", "done"))
+                next_reward = td.get(("next", "reward_raw"))[done]
+                next_kl = td.get(("next", "reward_kl"))[done]
+                rollout_rewards.append(next_reward.mean().cpu().item())
+                rollout_kl.append(next_kl.mean().cpu().item())
+            rollout_reward = torch.tensor(rollout_rewards).mean().cpu().item()
+            rollout_kl_reward = torch.tensor(rollout_kl).mean().cpu().item()
+            # recover true kl
+            rollout_kl = -rollout_kl_reward / kl_controller.value
+            kl_controller.update(rollout_kl, num_rollouts_per_epoch / batch_size)
+
+            # FIXME: THIS PPO CYCLE WAS DIFFERENT wrt trlx. @tcbegley please double check
+            # they sample batch_size from rb and then do minibatches ppo_batch_size within
+            if it % log_interval == 0:
+                val_reward_logger.info(
+                    f"TRAIN: {it=}: {rollout_reward=:.4f} {rollout_kl_reward=:.4f} {rollout_kl=:.4f}"
+                )
+                wandb.log(
+                    {
+                        "rollout_reward": rollout_reward,
+                        "rollout_kl_reward": rollout_kl_reward,
+                        "rollout_kl": rollout_kl,
+                    },
+                    step=it,
+                )
+                pbar.set_description(f"TRAIN: {it=}: {rollout_reward=:.4f}")
+
+            for batch in rb:
+                rb_ppo.empty()
+                rb_ppo.extend(batch)
+                for _ in range(ppo_num_epochs):  # PPO epochs
+                    optimizer.zero_grad()
+                    # why don't we optimize at each step? Is accumulating grads better?
+                    # usually more small steps is better than a giant one
```

Comment:
remove comment
by the way: the reason why ppl decompose a single minibatch in multiple mini minibatches is simply for memory constrains and because PPO in rlhf requires big batch size
If a singel minibatch fits in memory we can perfectly do that